
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Tutorial &#8212; Chorus 1.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Install" href="install.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<p>In this tutorial, we will build oligo porbe set for Arabidopsis genome.</p>
<div class="section" id="using-docker-terminal-version">
<h2>Using Docker Terminal Version<a class="headerlink" href="#using-docker-terminal-version" title="Permalink to this headline">¶</a></h2>
<div class="section" id="install">
<h3>Install<a class="headerlink" href="#install" title="Permalink to this headline">¶</a></h3>
<p>Install <a class="reference external" href="https://docs.docker.com/engine/installation/">Docker</a></p>
<p>Download Chorus:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ docker pull forrestzhang/docker-chorus
</pre></div>
</div>
<p>Parameter of Chorus:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-g GENOME, --genome GENOME
                        fasta format genome file
-i INPUT, --input INPUT
                        fasta format input file
-s SAVED, --save SAVED
                        result saved folder
-p PRIMER, --primer PRIMER
                        5&#39; labeled R primer
-t THREADS, --threads THREADS
                        threads number or how may cpu you wanna use
-l LENGTH, --length LENGTH
                        probe length
--homology HOMOLOGY   homology, from 50 to 100
-d DTM, --dtm DTM     dTm, from 0 to 37
</pre></div>
</div>
<p>Download Reference Genome file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ wget https://www.arabidopsis.org/download_files/Genes/TAIR10_genome_release/TAIR10_chromosome_files/TAIR10_chr_all.fas

$ docker run -v <span class="nv">$PWD</span>:/home/chorus -e <span class="nv">CHORUS_USER</span><span class="o">=</span><span class="nv">$USER</span> -e <span class="nv">CHORUS_UID</span><span class="o">=</span><span class="nv">$UID</span> <span class="se">\</span>
  forrestzhang/docker-chorus -i TAIR10_chr_all.fas -g TAIR10_chr_all.fas -t <span class="m">12</span>
</pre></div>
</div>
<p>Please wait unit all precess done. There are some logs:</p>
<div class="highlight-log notranslate"><div class="highlight"><pre><span></span>forrest /home/chorus
use local user:  forrest
Adding group &#39;forrest&#39; (GID 1000) ...
Done.
Adding user &#39;forrest&#39; ...
Adding new user &#39;forrest&#39; (1000) with group &#39;forrest&#39; ...
Creating home directory &#39;/home/forrest&#39; ...
Copying files from &#39;/etc/skel&#39; ...
/home/chorus exists
2.2.3
########################################
bwa version: /opt/software/bwa/bwa 0.7.12-r1044
jellyfish version: /opt/software/jellyfish/bin/jellyfish 2.2.3
genome file: TAIR10_chr_all.fas
input file: TAIR10_chr_all.fas
5&#39; labeled R primer:
result output folder: /home/chorus/probes
threads number: 12
homology: 75
dtm: 10
########################################
...
...
14300000 / 14326857
14310000 / 14326857
14320000 / 14326857
Job finshed!!
</pre></div>
</div>
<p>When process done:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ ls -lt probes/
total <span class="m">1741428</span>
-rw-r--r-- <span class="m">1</span> root root  <span class="m">280927981</span> Aug <span class="m">24</span> <span class="m">17</span>:44 TAIR10_chr_all.fas_all.bed
-rw-r--r-- <span class="m">1</span> root root   <span class="m">62050561</span> Aug <span class="m">24</span> <span class="m">17</span>:44 TAIR10_chr_all.fas.bed
-rw-r--r-- <span class="m">1</span> root root         <span class="m">94</span> Aug <span class="m">24</span> <span class="m">17</span>:30 TAIR10_chr_all.fas.len
-rw-r--r-- <span class="m">1</span> root root <span class="m">1031512169</span> Aug <span class="m">24</span> <span class="m">17</span>:22 TAIR10_chr_all.fas_tmp_probe.fa
-rw-r--r-- <span class="m">1</span> root root   <span class="m">59833928</span> Aug <span class="m">24</span> <span class="m">17</span>:19 TAIR10_chr_all.fas.sa
-rw-r--r-- <span class="m">1</span> root root       <span class="m">7535</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.amb
-rw-r--r-- <span class="m">1</span> root root        <span class="m">682</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.ann
-rw-r--r-- <span class="m">1</span> root root   <span class="m">29916939</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.pac
-rw-r--r-- <span class="m">1</span> root root  <span class="m">119667836</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.bwt
-rw-r--r-- <span class="m">1</span> root root  <span class="m">121183059</span> Aug <span class="m">24</span> <span class="m">17</span>:17 TAIR10_chr_all.fas
-rw-r--r-- <span class="m">1</span> root root   <span class="m">78102510</span> Aug <span class="m">24</span> <span class="m">17</span>:17 TAIR10_chr_all.fas_17mer.jf
</pre></div>
</div>
<p><em>TAIR10_chr_all.fas.bed</em> is the probe file.</p>
<div class="highlight-log notranslate"><div class="highlight"><pre><span></span>$ more probes/TAIR10_chr_all.fas.bed
1           52      96      TCCCTAAATCTTTAAATCCTACATCCATGAATCCCTAAATACCTA
1           211     255     TTTGAGGTCAATACAAATCCTATTTCTTGTGGTTTTCTTTCCTTC
1           346     390     CCTTAGGGTTGGTTTATCTCAAGAATCTTATTAATTGTTTGGACT
1           426     470     TTTGTGGAAATGTTTGTTCTATCAATTTATCTTTTGTGGGAAAAT
1           496     540     TCTTCGTTGTTGTTACGCTTGTCATCTCATCTCTCAATGATATGG
1           551     595     TAGCATTTATTCTGAAGTTCTTCTGCTTGATGATTTTATCCTTAG
</pre></div>
</div>
<p>There are four columns in a row, first column is chromosome name, second is oligo start site, third is oligo end site, last one is oligo probe sequence. You can use excel or text editor to open this file.</p>
</div>
</div>
<div class="section" id="using-manually-install-version">
<h2>Using Manually Install Version<a class="headerlink" href="#using-manually-install-version" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Install<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><a class="reference external" href="install.html#ubuntu-14-04-terminal">Manually Install</a></li>
</ul>
</div>
<div class="section" id="run-in-terminal">
<h3>Run In Terminal<a class="headerlink" href="#run-in-terminal" title="Permalink to this headline">¶</a></h3>
<p>Make a project folder</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> ~
$ mkdir sampleproject
$ <span class="nb">cd</span> sampleproject
</pre></div>
</div>
<p>Download reference genome</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ wget https://www.arabidopsis.org/download_files/Genes/TAIR10_genome_release/TAIR10_chromosome_files/TAIR10_chr_all.fas
</pre></div>
</div>
<p>Test chorus software</p>
<div class="highlight-bassh notranslate"><div class="highlight"><pre><span></span>$ python3 /opt/software/Chorus/Chorus.py -h

    usage: Chorus [-h] [--version] [-j JELLYFISH] [-b BWA] -g GENOME -i INPUT
            [-s SAVED] [-p PRIMER] [-t THREADS] [-l LENGTH]
            [--homology HOMOLOGY] [-d DTM] [--step STEP] [--docker DOCKER]

    Chorus Software for Oligo FISH probe design

    optional arguments:
    -h, --help            show this help message and exit
    --version             show program&#39;s version number and exit
    -j JELLYFISH, --jellyfish JELLYFISH
                            jellyfish path
    -b BWA, --bwa BWA     bwa path
    -g GENOME, --genome GENOME
                            fasta format genome file
    -i INPUT, --input INPUT
                            fasta format input file
    -s SAVED, --save SAVED
                            result saved folder
    -p PRIMER, --primer PRIMER
                            5&#39; labeled R primer
    -t THREADS, --threads THREADS
                            threads number or how may cpu you wanna use
    -l LENGTH, --length LENGTH
                            probe length
    --homology HOMOLOGY   homology, from 50 to 100
    -d DTM, --dtm DTM     dTm, from 0 to 37
    --step STEP           step length, min=1
    --docker DOCKER
</pre></div>
</div>
<p>Run chorus software</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python3 /opt/software/Chorus/Chorus.py -i TAIR10_chr_all.fas -g TAIR10_chr_all.fas -t <span class="m">12</span> <span class="se">\</span>
  -j /opt/software/jellyfish/bin/jellyfish -b /opt/software/bwa/bwa -s sample
</pre></div>
</div>
<p>When job finish, the oligo probes will output to ‘sample’ folder</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> sample
$ ls -lt *

    total <span class="m">1741428</span>
    -rw-r--r-- <span class="m">1</span> root root  <span class="m">280927981</span> Aug <span class="m">24</span> <span class="m">17</span>:44 TAIR10_chr_all.fas_all.bed
    -rw-r--r-- <span class="m">1</span> root root   <span class="m">62050561</span> Aug <span class="m">24</span> <span class="m">17</span>:44 TAIR10_chr_all.fas.bed
    -rw-r--r-- <span class="m">1</span> root root         <span class="m">94</span> Aug <span class="m">24</span> <span class="m">17</span>:30 TAIR10_chr_all.fas.len
    -rw-r--r-- <span class="m">1</span> root root <span class="m">1031512169</span> Aug <span class="m">24</span> <span class="m">17</span>:22 TAIR10_chr_all.fas_tmp_probe.fa
    -rw-r--r-- <span class="m">1</span> root root   <span class="m">59833928</span> Aug <span class="m">24</span> <span class="m">17</span>:19 TAIR10_chr_all.fas.sa
    -rw-r--r-- <span class="m">1</span> root root       <span class="m">7535</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.amb
    -rw-r--r-- <span class="m">1</span> root root        <span class="m">682</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.ann
    -rw-r--r-- <span class="m">1</span> root root   <span class="m">29916939</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.pac
    -rw-r--r-- <span class="m">1</span> root root  <span class="m">119667836</span> Aug <span class="m">24</span> <span class="m">17</span>:18 TAIR10_chr_all.fas.bwt
    -rw-r--r-- <span class="m">1</span> root root  <span class="m">121183059</span> Aug <span class="m">24</span> <span class="m">17</span>:17 TAIR10_chr_all.fas
    -rw-r--r-- <span class="m">1</span> root root   <span class="m">78102510</span> Aug <span class="m">24</span> <span class="m">17</span>:17 TAIR10_chr_all.fas_17mer.jf
</pre></div>
</div>
<p><em>TAIR10_chr_all.fas.bed</em> is the probe file.</p>
<div class="highlight-log notranslate"><div class="highlight"><pre><span></span>$ more probes/TAIR10_chr_all.fas.bed
1           52      96      TCCCTAAATCTTTAAATCCTACATCCATGAATCCCTAAATACCTA
1           211     255     TTTGAGGTCAATACAAATCCTATTTCTTGTGGTTTTCTTTCCTTC
1           346     390     CCTTAGGGTTGGTTTATCTCAAGAATCTTATTAATTGTTTGGACT
1           426     470     TTTGTGGAAATGTTTGTTCTATCAATTTATCTTTTGTGGGAAAAT
1           496     540     TCTTCGTTGTTGTTACGCTTGTCATCTCATCTCTCAATGATATGG
1           551     595     TAGCATTTATTCTGAAGTTCTTCTGCTTGATGATTTTATCCTTAG
</pre></div>
</div>
<p>There are four columns in a row, first column is chromosome name, second is oligo start site, third is oligo end site, last one is oligo probe sequence. You can use excel or text editor to open this file.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial</a><ul>
<li><a class="reference internal" href="#using-docker-terminal-version">Using Docker Terminal Version</a><ul>
<li><a class="reference internal" href="#install">Install</a></li>
</ul>
</li>
<li><a class="reference internal" href="#using-manually-install-version">Using Manually Install Version</a><ul>
<li><a class="reference internal" href="#id1">Install</a></li>
<li><a class="reference internal" href="#run-in-terminal">Run In Terminal</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="install.html" title="previous chapter">Install</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/tutorial.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Tao Zhang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/tutorial.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>